### PR TITLE
support 0 shares in PVSS

### PR DESCRIPTION
**Pull request by zjma**
## Description

Add support of players with 0 shares in weighted Das PVSS, so we can support an infallible rounding algorithm.

## Type of Change
- [x] New feature

## Which Components or Systems Does This Change Impact?
No system-level impact.

## How Has This Been Tested?
UTs

## Key Areas to Review
None.

## Checklist
- [x] I have read and followed the [CONTRIBUTING](https://github.com/aptos-labs/aptos-core/blob/main/CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I identified and added all stakeholders and component owners affected by this change as reviewers
- [x] I tested both happy and unhappy path of the functionality
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

**Review by alinush — CHANGES_REQUESTED** [review]
First of all, please fill in the PR template :)

File: crates/aptos-dkg/src/pvss/weighted/weighted_config.rs [hunk 1]
```diff
@@ -130,7 +124,6 @@ impl WeightedConfig {
     ///
     /// Returns the index of this player's share in the vector of shares, or None if out of bounds.
     pub fn get_share_index(&self, i: usize, j: usize) -> Option<usize> {
-        assert_lt!(i, self.tc.n);
```
Comment (alinush) [inline crates/aptos-dkg/src/pvss/weighted/weighted_config.rs:133]: Why remove this?

Was this wrong actually? `self.tc.n` is the total # of shares, not players...

**Review by alinush — CHANGES_REQUESTED** [review]
Can you add tests in `aptos-dkg` for the 0 shares?

File: crates/aptos-dkg/src/pvss/weighted/weighted_config.rs (hunk 1 above)
Comment (zjma) [reply crates/aptos-dkg/src/pvss/weighted/weighted_config.rs]: yup, if everyone gets 1 share except that validator 0 gets 0 shares, `self.tc.n == n-1` and `i` can be `n-1` too.

File: crates/aptos-dkg/src/pvss/das/weighted_protocol.rs [hunk 2]
```diff
@@ -538,3 +538,125 @@ impl Transcript {
         }
     }
 }
+
```
Comment (alinush) [inline crates/aptos-dkg/src/pvss/das/weighted_protocol.rs:541]: Why are you doing it here only for `das` instead of in `tests/pvss.rs`?

Comment (zjma) [reply crates/aptos-dkg/src/pvss/das/weighted_protocol.rs]: oops, fixed

**Review by alinush — APPROVED** [review]

**Review by alinush — COMMENTED** [review]
@zjma can you fill in the **full** PR template, please?

A lot of items were deleted. Let's do things by the book, so we can tell if we are being thorough enough.

If those items were irrelevant, instead of deleting them, just write under them why they are irrelevant.

File: crates/aptos-dkg/src/pvss/weighted/weighted_config.rs (hunk 1 above)
Comment (danielxiangzl) [reply crates/aptos-dkg/src/pvss/weighted/weighted_config.rs:133]: then should we assert `assert_le`?

Comment (zjma) [reply crates/aptos-dkg/src/pvss/weighted/weighted_config.rs:133]: we are allowing num_players > num_shares, so we shouldn't.

File: crates/aptos-dkg/src/pvss/test_utils.rs [hunk 3]
```diff
@@ -164,6 +164,9 @@ pub fn get_weighted_configs_for_testing() -> Vec<WeightedConfig> {
     // 50-out-of-100, weights [11, 13, 9, 10, 12, 8, 7, 14, 10, 6]
     wcs.push(WeightedConfig::new(50, vec![11, 13, 9, 10, 12, 8, 7, 14, 10, 6]).unwrap());
 
+    // 7-out-of-15, weights [0, 0, 0, 2, 2, 2, 0, 0, 0, 3, 3, 3, 0, 0, 0]
+    wcs.push(WeightedConfig::new(7, vec![0, 0, 0, 2, 2, 2, 0, 0, 0, 3, 3, 3, 0, 0, 0]).unwrap());
+
```
Comment (danielxiangzl) [inline crates/aptos-dkg/src/pvss/test_utils.rs:169]: Add more corner test cases? like all 0's, 0 out of X, X out of X?

**Review by danielxiangzl — APPROVED** [review]

**Pull request merged**